flink-libraries/flink-state-processing-api/src/main/java/org/apache/flink/state/api/WritableSavepoint.java

```diff
@@ -0,0 +1,129 @@
+import org.apache.flink.api.java.DataSet;
+import org.apache.flink.core.fs.Path;
+import org.apache.flink.runtime.checkpoint.OperatorState;
+import org.apache.flink.runtime.state.StateBackend;
+import org.apache.flink.state.api.output.OperatorStateReducer;
+import org.apache.flink.state.api.output.OperatorSubtaskStateReducer;
+import org.apache.flink.state.api.output.SavepointOutputFormat;
+import org.apache.flink.state.api.runtime.metadata.SavepointMetadata;
+
+import javax.annotation.Nullable;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Any savepoint that can be written to from a batch context.
+ * @param <F> The implementation type.
+ */
+@PublicEvolving
+@SuppressWarnings("WeakerAccess")
+public abstract class WritableSavepoint<F extends WritableSavepoint> {
+
+	protected final Map<String, BootstrapTransformation> transformations;
+
+	protected final List<String> droppedOperators;
+
+	WritableSavepoint() {
+		this.transformations  = new HashMap<>();
+		this.droppedOperators = new ArrayList<>();
+	}
+
+	/**
+	 * Drop an existing operator from the savepoint.
+	 * @param uid The uid of the operator.
+	 * @return A modified savepoint.
+	 */
+	@SuppressWarnings("unchecked")
+	public F removeOperator(String uid) {
+		droppedOperators.add(uid);
+		transformations.remove(uid);
+		return (F) this;
+	}
+
+	/**
+	 * Adds a new operator to the savepoint.
+	 * @param uid The uid of the operator.
+	 * @param transformation The operator to be included.
+	 * @return The modified savepoint.
+	 */
+	@SuppressWarnings("unchecked")
+	public <T> F withOperator(String uid, BootstrapTransformation<T> transformation) {
+		if (transformations.containsKey(uid)) {
+			throw new IllegalArgumentException("The savepoint already contains uid " + uid + ". All uid's must be unique");
+		}
+
+		transformations.put(uid, transformation);
+		return (F) this;
+	}
+
+	/**
+	 * Write out a new or updated savepoint.
+	 * @param path The path to where the savepoint should be written.
+	 */
+	public abstract void write(String path);
+
+	protected static void write(
+		Path savepointPath,
+		Map<String, BootstrapTransformation> transformations,
+		StateBackend stateBackend,
+		SavepointMetadata metadata,
+		@Nullable DataSet<OperatorState> existingOperators) {
+
+		DataSet<OperatorState> newOperatorStates = transformations
+			.entrySet()
+			.stream()
+			.map(entry -> getOperatorStates(savepointPath, entry.getKey(), stateBackend, entry.getValue(), metadata))
+			.reduce(DataSet::union)
+			.orElseThrow(() -> new IllegalStateException("Savepoint's must contain at least one operator"));
+
+		DataSet<OperatorState> finalOperatorStates;
+		if (existingOperators == null) {
+			finalOperatorStates = newOperatorStates;
+		} else {
+			finalOperatorStates = newOperatorStates.union(existingOperators);
+		}
+
+		finalOperatorStates
+			.reduceGroup(new OperatorStateReducer(metadata))
```

Comment:
Following my comment regarding how we should use the `savepointMetadata` to maintain an index of `Map<String(uid), Either<OperatorState, BootstrapTransformation>>`, I think this reduce can even be avoided.

The main logic for `write` would essentially become:
1. For each entry in the index, write the operator state data. If it is an `OperatorState`, then we simply re-write the data. If it is a `BootstrapTransformation`, then we perform the logic in `transformation.getOperatorSubtaskStates(...)`.
2. By now, all values in the `SavepointMetadata` index should be `OperatorState`s, and no longer `BootstrapTransformation`s. Simply output the `SavepointMetadata` using `SavepointOutputFormat`. The output format internall transforms our `SavepointMetadata` back to Flink's `Savepoint`, just for the sake of API usage.